### PR TITLE
[Carry 2535] rootless: support detach-netns mode

### LINKED ISSUE #86: rootless: provide workaround for `nerdctl pull 127.0.0.1:5000/foo`
`nerdctl pull 127.0.0.1:5000/foo` with rootless doesn't work as expected because `pull` occurs in the child namespace.

Needs something like `socat   exec:'containerd-rootless-setuptool.sh nsenter socat "tcp-listen:5000,reuseaddr,fork" "stdio"' tcp-connect:127.0.0.1:5000` (WIP: this exits on HTTP error)

### LINKED ISSUE #814: rootless: run the puller/pusher in the host netns (but still in the userns)
With the upcoming bypass4netns integration, rootless containers no longer needs to be slowed down with slirp, however, the puller and the pusher are still slowed down.

We should now look into possibility of running the puller/pusher in the host netns to avoid the slirp overhead.

### PR DESCRIPTION
**Pull request by AkihiroSuda**
When RootlessKit v2.0 (rootless-containers/rootlesskit#379) is installed, `containerd-rootless.sh` launches it with `--detach-netns` so as to run the daemon in the host network namespace.

This will enable:
- Accelerated (and deflaked) `nerdctl pull`, `nerdctl push`, `nerdctl build`, etc
- Proper support for `nerdctl pull 127.0.0.1:.../...`
- Proper support for `nerdctl run --net=host`

Replaces Fahed Dorgaa's PR #2535
Fixes #814
Fixes #86

- - -

> [!NOTE]
> Rootless containerd recognizes the following environment variables to configure the behavior of [RootlessKit](https://github.com/rootless-containers/rootlesskit):
> * ...
> * `CONTAINERD_ROOTLESS_ROOTLESSKIT_DETACH_NETNS=(auto|true|false)`: whether to launch rootlesskit with the "detach-netns" mode.
>   Defaults to "auto", which is resolved to "true" if RootlessKit >= 2.0 is installed.
>   The "detached-netns" mode accelerates `nerdctl (pull|push|build)` and enables `nerdctl run --net=host`,
>   however, there is a relatively minor drawback with BuildKit prior to v0.13:
>   the host loopback IP address (127.0.0.1) is exposed to Dockerfile's "RUN" instructions during `nerdctl build` (not `nerdctl run`).
>  The drawback is fixed in BuildKit v0.13. Upgrading from a prior version of BuildKit needs removing the old systemd unit:
> `containerd-rootless-setuptool.sh uninstall-buildkit && rm -f ~/.config/buildkit/buildkitd.toml`
> 
> To set these variables, create `~/.config/systemd/user/containerd.service.d/override.conf` as follows:
> ```ini
> [Service]
> Environment=CONTAINERD_ROOTLESS_ROOTLESSKIT_DETACH_NETNS="false"
> ```
> 
> And then run the following commands:
> ```bash
> systemctl --user daemon-reload
> systemctl --user restart containerd
> ```

https://github.com/containerd/nerdctl/blob/main/docs/rootless.md

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
```
test target: "nerdctl"
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuild (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildBaseImage (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildFromContainerd (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildWithDockerfile (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildWithBuildArg (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildMultipleTags (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildWithContainerfile (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildWithDockerFileAndContainerfile (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildNoTag (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildSourceDateEpoch (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildNetwork (0.02s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestBuildNetworkShellCompletion (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestBuilderDebug (6.79s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestCompletion (3.49s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestComposeBuild (0.02s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfig (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithPrintService (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithPrintServiceHash (0.11s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithMultipleFile (0.11s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithComposeFileEnv (0.11s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithEnvFile (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCopy (0.47s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCreate (0.40s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCreateDependency (0.32s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCreatePull (6.88s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCreateBuild (0.02s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeDownRemoveUsedNetwork (0.63s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeDownRemoveOrphans (0.67s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExec (0.52s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExecWithEnv (0.41s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExecWithUser (0.81s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExecTTY (0.76s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExecWithIndex (2.90s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeImages (14.56s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeImagesJson (1.07s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeKill (4.00s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestComposePauseAndUnpause (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePort (0.46s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePortFailure (0.47s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePs (4.32s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePsJSON (2.26s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePullWithService (0.66s)
=== RUN   TestComposeRestart
    compose_restart_linux_test.go:61: projectName="nerdctl-compose-test1674241425"
    compose_restart_linux_test.go:70: assertion failed: error is not nil: expected stdout to contain any of "Up|running": NAME    IMAGE    COMMAND    SERVICE    STATUS    PORTS
        
--- FAIL: TestComposeRestart (11.20s)
FAIL github.com/containerd/nerdctl/cmd/nerdctl.TestComposeRestart (11.20s)
```

https://github.com/containerd/nerdctl/actions/runs/7374442145/job/20064826128?pr=2723

**Comment by AkihiroSuda** [comment]
https://github.com/containerd/nerdctl/actions/runs/7380993927/job/20079027998?pr=2723

```
test target: "nerdctl"
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuild (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildBaseImage (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildFromContainerd (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildWithDockerfile (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildWithBuildArg (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildMultipleTags (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildWithContainerfile (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildWithDockerFileAndContainerfile (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildNoTag (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildSourceDateEpoch (0.02s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestBuildNetwork (0.02s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestBuildNetworkShellCompletion (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestBuilderDebug (1.17s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestCompletion (2.34s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestComposeBuild (0.02s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfig (0.04s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithPrintService (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithPrintServiceHash (0.10s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithMultipleFile (0.11s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithComposeFileEnv (0.10s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeConfigWithEnvFile (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCopy (0.47s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCreate (0.40s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCreateDependency (0.31s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCreatePull (3.29s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestComposeCreateBuild (0.02s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeDownRemoveUsedNetwork (0.65s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeDownRemoveOrphans (0.65s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExec (0.55s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExecWithEnv (0.41s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExecWithUser (0.83s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExecTTY (0.76s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeExecWithIndex (2.83s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeImages (14.54s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeImagesJson (0.95s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeKill (3.95s)
SKIP github.com/containerd/nerdctl/cmd/nerdctl.TestComposePauseAndUnpause (0.03s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePort (0.46s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePortFailure (0.46s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePs (4.27s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePsJSON (2.18s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposePullWithService (0.55s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeRestart (12.96s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeRemove (11.71s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeRun (0.38s)
PASS github.com/containerd/nerdctl/cmd/nerdctl.TestComposeRunWithRM (0.42s)
=== RUN   TestComposeRunWithServicePorts
    compose_run_linux_test.go:121: projectName="nerdctl-compose-test2912223502"
    compose_run_linux_test.go:150: (retry 0)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": dial tcp 127.0.0.1:8080: connect: connection refused
    compose_run_linux_test.go:150: (retry 1)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 2)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 3)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 4)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 5)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 6)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 7)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 8)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 9)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 10)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 11)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 12)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 13)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 14)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 15)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 16)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 17)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 18)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 19)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 20)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 21)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 22)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 23)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 24)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 25)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 26)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 27)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 28)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:150: (retry 29)
    compose_run_linux_test.go:156: error after 10 attempts: Get "http://127.0.0.1:8080": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    compose_run_linux_test.go:160: nginx is not working
--- FAIL: TestComposeRunWithServicePorts (990.76s)
FAIL github.com/containerd/nerdctl/cmd/nerdctl.TestComposeRunWithServicePorts (990.76s)
```

**Comment by AkihiroSuda** [comment]
https://github.com/containerd/nerdctl/actions/runs/7406193082/job/20150258934?pr=2723
```
=== RUN   TestHostsFileMounts
    container_run_network_linux_test.go:507: assertion failed: res.ExitCode is not exitCode: sh: can't create /etc/hosts: Permission denied
        
--- FAIL: TestHostsFileMounts (0.72s)
```

File: extras/rootless/containerd-rootless.sh [hunk 1]
```diff
@@ -38,6 +38,14 @@
 # * CONTAINERD_ROOTLESS_ROOTLESSKIT_PORT_DRIVER=(builtin|slirp4netns): the rootlesskit port driver. Defaults to "builtin".
 # * CONTAINERD_ROOTLESS_ROOTLESSKIT_SLIRP4NETNS_SANDBOX=(auto|true|false): whether to protect slirp4netns with a dedicated mount namespace. Defaults to "auto".
 # * CONTAINERD_ROOTLESS_ROOTLESSKIT_SLIRP4NETNS_SECCOMP=(auto|true|false): whether to protect slirp4netns with seccomp. Defaults to "auto".
+# * CONTAINERD_ROOTLESS_ROOTLESSKIT_DETACH_NETNS=(auto|true|false): whether to launch rootlesskit with the "detach-netns" mode.
+#   Defaults to "auto", which is resolved to "true" if RootlessKit >= 2.0 is installed.
+#   The "detached-netns" mode accelerates `nerdctl (pull|push|build)` and enables `nerdctl run --net=host`,
+#   however, there is a relatively minor drawback with the current version of BuildKit:
+#   the host loopback IP address (127.0.0.1) is exposed to Dockerfile's "RUN" instructions during `nerdctl build` (not `nerdctl run`).
+#   If you want to hide 127.0.0.1 from "RUN" instructions, you should set CONTAINERD_ROOTLESS_ROOTLESSKIT_DETACH_NETNS to "false".
```
Comment (AkihiroSuda) [inline extras/rootless/containerd-rootless.sh:46]: @containerd/nerdctl-maintainers 

This is kinda breaking change.
Let me know we should release this as "nerdctl v2.0" (ahead of containerd v2.0), or we can just stick to "nerdctl  v1.8".
I'm leaning toward to the former one.

Comment (ktock) [reply extras/rootless/containerd-rootless.sh:46]: > nerdctl v2.0

SGTM

**Review by ktock — APPROVED** [review]

File: docs/rootless.md [hunk 2]
```diff
@@ -130,6 +130,35 @@ $ nerdctl run -it --rm -p 8080:80 --label nerdctl/bypass4netns=true alpine
...
+* `CONTAINERD_ROOTLESS_ROOTLESSKIT_PORT_DRIVER=(builtin|slirp4netns)`: the rootlesskit port driver. Defaults to "builtin".
+* `CONTAINERD_ROOTLESS_ROOTLESSKIT_SLIRP4NETNS_SANDBOX=(auto|true|false)`: whether to protect slirp4netns with a dedicated mount namespace. Defaults to "auto".
+* `CONTAINERD_ROOTLESS_ROOTLESSKIT_SLIRP4NETNS_SECCOMP=(auto|true|false)`: whether to protect slirp4netns with seccomp. Defaults to "auto".
+* `CONTAINERD_ROOTLESS_ROOTLESSKIT_DETACH_NETNS=(auto|true|false)`: whether to launch rootlesskit with the "detach-netns" mode.
+  Defaults to "auto", which is resolved to "true" if RootlessKit >= 2.0 is installed.
+  The "detached-netns" mode accelerates `nerdctl (pull|push|build)` and enables `nerdctl run --net=host`,
+  however, there is a relatively minor drawback with the current version of BuildKit (planned to be fixed in [BuildKit v0.13](https://github.com/moby/buildkit/pull/4546)):
+  the host loopback IP address (127.0.0.1) is exposed to Dockerfile's "RUN" instructions during `nerdctl build` (not `nerdctl run`).
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply docs/rootless.md:147]: What do you mean?

File: extras/rootless/containerd-rootless-setuptool.sh [hunk 3]
```diff
@@ -152,8 +152,13 @@ propagate_env_from() {
 cmd_entrypoint_nsenter() {
 	# No need to call init()
 	pid=$(cat "$XDG_RUNTIME_DIR/containerd-rootless/child_pid")
+	n=""
+	# If RootlessKit is running with `--detach-netns` mode, we do NOT enter the detached netns here
+	if [ ! -e "$XDG_RUNTIME_DIR/containerd-rootless/netns" ]; then
+		n="-n"
+	fi
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply extras/rootless/containerd-rootless-setuptool.sh:159]: They can just run /usr/bin/nsenter -n$XDG_RUNTIME_DIR/netns

File: docs/rootless.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (fahedouch) [reply docs/rootless.md:147]: I don't  understand why invoking `nerdctl build` here  : Does `nerdctl build` prevent the loopback address from being exposed for `nerdctl run`?

File: pkg/cmd/container/create.go [hunk 4]
```diff
@@ -418,6 +419,29 @@ func GenerateLogURI(dataStore string) (*url.URL, error) {
...
+			//     - args: {"--data-root=/foo", "internal", "oci-hook"}
+			//   - New:
+			//     - cmd:  "/usr/bin/nsenter"
+			//     - args: {"-n/run/user/1000/containerd-rootless/netns", "-F", "--", "/usr/local/bin/nerdctl", "--data-root=/foo", "internal", "oci-hook"}
+			oldCmd, oldArgs := cmd, args
+			cmd, err = exec.LookPath("nsenter")
+			if err != nil {
+				return nil, err
+			}
+			args = append([]string{"-n" + detachedNetNS, "-F", "--", oldCmd}, oldArgs...)
```
Comment (fahedouch) [inline pkg/cmd/container/create.go:441]: So, in the end, we decided not to create a nested network namespace for each container inside a detached network namespace ?

Comment (AkihiroSuda) [reply pkg/cmd/container/create.go:441]: The OCI hook creates nested namespaces

File: docs/rootless.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply docs/rootless.md:147]: `nerdctl run` and `RUN` instructions in `nerdctl build` have different code paths.
One does not affect another.

File: pkg/netutil/subnet/subnet.go [hunk 5]
```diff
@@ -19,11 +19,17 @@ package subnet
...
+	"github.com/containerd/nerdctl/v2/pkg/rootlessutil"
 )
 
 func GetLiveNetworkSubnets() ([]*net.IPNet, error) {
-	addrs, err := net.InterfaceAddrs()
-	if err != nil {
+	var addrs []net.Addr
+	if err := rootlessutil.WithDetachedNetNSIfAny(func() error {
+		var err2 error
+		addrs, err2 = net.InterfaceAddrs()
```
Comment (fahedouch) [inline pkg/netutil/subnet/subnet.go:30]: Ah, I think this is what I was missing in my PR; I wasn't pointing to the right IP addresses, right? It's not an issue with `runc` as we discussed in Chicago.

File: docs/rootless.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (fahedouch) [reply docs/rootless.md:147]: nvm!

File: pkg/cmd/container/create.go (hunk 4 above)
Comment (fahedouch) [reply pkg/cmd/container/create.go:441]: Yes 👍

**Comment by fahedouch** [comment]
thanks @AkihiroSuda 

LGTM overall, just this small [typo](https://github.com/containerd/nerdctl/pull/2723#discussion_r1457010556)

Thanks again for taking the time to address this feature

**Review by fahedouch — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Will rebase when BuildKit v0.13 beta2 is released

**Review by ktock — APPROVED** [review]

File: extras/rootless/containerd-rootless.sh [hunk 6]
```diff
@@ -107,6 +117,25 @@ if [ -z $_CONTAINERD_ROOTLESS_CHILD ]; then
 			export _CONTAINERD_ROOTLESS_SELINUX
 		fi
 	fi
+
+	case "$CONTAINERD_ROOTLESS_ROOTLESSKIT_DETACH_NETNS" in
+	auto)
+		if rootlesskit --help | grep -qw -- "--detach-netns"; then
+			CONTAINERD_ROOTLESS_ROOTLESSKIT_FLAGS=--detach-netns $CONTAINERD_ROOTLESS_ROOTLESSKIT_FLAGS
```
Comment (apostasie) [inline extras/rootless/containerd-rootless.sh:124]: I do not think this is doing what you think.

What will happen instead is:

```
VAR=VALUE command
```

Will fix.

Comment (AkihiroSuda) [reply extras/rootless/containerd-rootless.sh:124]: Fixed in https://github.com/containerd/nerdctl/pull/3214 (thanks)

**Pull request merged**